ql/src/java/org/apache/hadoop/hive/ql/history/QueryCompletedEventScriber.java

```diff
@@ -0,0 +1,296 @@
+      thriftCounterInfo.setCounters(ent.getValue().getCounters().toString());
+      thriftCounterInfo.setNumberMappers(ent.getValue().getNumMap());
+      thriftCounterInfo.setNumberReducers(ent.getValue().getNumReduce());
+      thriftCounterInfo.setTaskNumbers(ent.getValue().getTaskNumbers());
+      thriftMapReduceInfo.put(key, thriftCounterInfo);
+    }
+  }
+
+  private static void setStageList(List<StageInfo> thriftStageList, List<Stage> stageList) {
+    if (stageList == null) {
+      return;
+    }
+    for (Stage stage : stageList) {
+      StageInfo stageEnt = new StageInfo();
+      stageEnt.setStageId(stage.getStageId());
+      stageEnt.setStageType(stage.getStageType().toString());
+
+      stageEnt.setStageAttributes(new HashMap<>());
+      setMapValForString(stageEnt.getStageAttributes(), stage.getStageAttributes());
+
+      stageEnt.setStageCounters(new HashMap<>());
+      setMapValForLong(stageEnt.getStageCounters(), stage.getStageCounters());
+
+      stageEnt.setTaskList(new ArrayList<>());
+      setTaskList(stageEnt.getTaskList(), stage.getTaskList());
+
+      stageEnt.setDone(stage.isDone());
+      stageEnt.setStarted(stage.isStarted());
+      thriftStageList.add(stageEnt);
+    }
+  }
+
+  private static void setTaskList(List<TaskInfo> thriftTaskList, List<Task> taskList) {
+    if (taskList == null) {
+      return;
+    }
+    for (Task task : taskList) {
+      TaskInfo thriftTaskListEnt = new TaskInfo();
+      thriftTaskListEnt.setTaskId(task.getTaskId());
+      thriftTaskListEnt.setTaskType(task.getTaskType().toString());
+      thriftTaskListEnt.setDone(task.isDone());
+      thriftTaskListEnt.setStarted(task.isStarted());
+
+      thriftTaskListEnt.setTaskAttributes(new HashMap<>());
+      setMapValForString(thriftTaskListEnt.getTaskAttributes(), task.getTaskAttributes());
+
+      thriftTaskListEnt.setTaskCounters(new HashMap<>());
+      setMapValForLong(thriftTaskListEnt.getTaskCounters(), task.getTaskCounters());
+
+      thriftTaskListEnt.setOperatorGraph(new GraphInfo());
+      setOperatorGraph(thriftTaskListEnt.getOperatorGraph(), task.getOperatorGraph());
+
+      thriftTaskListEnt.setOperatorList(new ArrayList<>());
+      setOperatorList(thriftTaskListEnt.getOperatorList(), task.getOperatorList());
+
+      thriftTaskList.add(thriftTaskListEnt);
+    }
+  }
+
+  private static void setTaskProgress(List<TaskDetailInfo> thriftTaskProgress, ArrayList<QueryStats.taskDetail> taskProgress) {
+    if (taskProgress == null) {
+      return;
+    }
+    for (QueryStats.taskDetail tDetail : taskProgress) {
+      TaskDetailInfo thriftProgressInfo = new TaskDetailInfo();
+      thriftProgressInfo.setTimeStamp(tDetail.getTimeStamp());
+      thriftProgressInfo.setProgress(tDetail.getProgress());
+      thriftTaskProgress.add(thriftProgressInfo);
+    }
+  }
+
+  private static void setMapValForString(Map<String, String> thriftConfigs, Map<String, String> configs) {
```

Comment:
You should be able to use a constructor (`HashMap(Map<? extends K,? extends V> m)`) or `putAll(Map<? extends K,? extends V> m)` to complete this.

Reply:
I am thinking that deepcopy may be safer here. pullAll doesn't do deepcopy. Since the current code has been running ok on prod. Can we do a merge for this stable version for now?

Reply:
consider using `ImmutableMap.copyOf(m)`